src/main/java/com/vmware/vipclient/i18n/messages/service/ProductService.java

```diff
@@ -77,18 +73,31 @@ public ProductService(BaseDTO dto) {
      *
      * @return list of locales of the product specified in the dto object
      */
-    public List<String> getSupportedLocales(){
-        List<String> locales = null;
+    public Set<String> getSupportedLocales(){
+        FormattingCacheService cacheService = new FormattingCacheService();
+        Map<String, String> supportedLanguages = cacheService.getSupportedLanguages(dto);
+        if (supportedLanguages != null && !supportedLanguages.isEmpty())
+            return supportedLanguages.keySet();
+
+        Set<String> locales = new HashSet<>();
         Iterator<DataSourceEnum> msgSourceQueueIter = VIPCfg.getInstance().getMsgOriginsQueue().iterator();
-        while((locales == null || locales.isEmpty()) && msgSourceQueueIter.hasNext()){
+        while(locales.isEmpty() && msgSourceQueueIter.hasNext()){
             DataSourceEnum dataSource = msgSourceQueueIter.next();
             ProductOpt opt = dataSource.createProductOpt(dto);
-            locales = opt.getSupportedLocales();
+            locales.addAll(opt.getSupportedLocales());
             // If failed to get locales from the data source, log the error.
-            if (locales == null || locales.isEmpty()) {
+            if (locales.isEmpty()) {
                 logger.error(ConstantsMsg.GET_LOCALES_FAILED, dataSource.toString());
             }
         }
+
+        // Add list of supported locales to cache
+        Map<String, String> cacheMap = new HashMap<>();
+        for (String locale : locales) {
+            cacheMap.put(locale, "");
```

Comment:
I think you needn't convert locale list to locale map, just add locale list to cache is enough.

Reply:
FormattingCacheItem has a Map<String,String> property. How do you add a list instead?

Reply:
Get it, it make sense to convert it to a map for reusing cacheItem Object.